### PR TITLE
download with powershell 2.0 for windows

### PR DESCRIPTION
**Pull request by musm**
Supported on all versions starting from windows 7

### CODE REVIEW

File: base/interactiveutil.jl [hunk 1]
```diff
@@ -571,11 +571,8 @@ end
 downloadcmd = nothing
 if is_windows()
     function download(url::AbstractString, filename::AbstractString)
-        res = ccall((:URLDownloadToFileW,:urlmon),stdcall,Cuint,
-                    (Ptr{Void},Cwstring,Cwstring,Cuint,Ptr{Void}),C_NULL,url,filename,0,C_NULL)
-        if res != 0
-            error("automatic download failed (error: $res): $url")
-        end
+        ps = "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe"
+        res = run(`$ps -NoProfile -Command "(New-Object System.Net.Webclient).DownloadFile(\"$url\", \"$filename\")"`)
```
Comment (vtjnash) [inline base/interactiveutil.jl]: `DownloadFile('$(replace(url, "'", "''"))', '$(replace(filename, "'", "''"))'`

**Comment by vtjnash** [comment]
Also, do we need to enable secure HTTPS support explicitly (https://github.com/JuliaPackaging/WinRPM.jl/issues/57#issuecomment-302778675)?

File: base/interactiveutil.jl (hunk 1 above)
Comment (musm) [inline base/interactiveutil.jl]: I admit I don't know what that does here, seem more complicated but does the same job?

Comment (vtjnash) [inline base/interactiveutil.jl]: it ensures that meta characters (`` ` ``, `$`, `'`) pass through unscathed

**Comment by musm** [comment]
I enabled https support. Yes it looks like we need to do it explicitly.
I updated the code to always enable https tls1.2 support.

File: base/interactiveutil.jl [hunk 2]
```diff
@@ -571,11 +571,10 @@ end
 downloadcmd = nothing
 if is_windows()
     function download(url::AbstractString, filename::AbstractString)
-        res = ccall((:URLDownloadToFileW,:urlmon),stdcall,Cuint,
-                    (Ptr{Void},Cwstring,Cwstring,Cuint,Ptr{Void}),C_NULL,url,filename,0,C_NULL)
-        if res != 0
-            error("automatic download failed (error: $res): $url")
-        end
+        ps = "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe"
```
Comment (musm) [inline base/interactiveutil.jl]: I believe powershell will always be in path so it may be best to remove the explicit location?

**Comment by musm** [comment]
refs https://github.com/JuliaLang/julia/issues/19526
https://github.com/JuliaLang/BinDeps.jl/issues/135

**Comment by musm** [comment]
bump?

File: base/interactiveutil.jl [hunk 3]
```diff
@@ -571,11 +571,10 @@ end
 downloadcmd = nothing
```
Comment (tkelman) [inline base/interactiveutil.jl]: may as well set this, if a command is being used

File: base/interactiveutil.jl (hunk 1 above)
Comment (tkelman) [inline base/interactiveutil.jl]: should have some tests for this then? likewise with the unix version

**Review by vtjnash — APPROVED** [review]

File: base/interactiveutil.jl (hunk 3 above)
Comment (musm) [inline base/interactiveutil.jl]: I think `downloadcmd` can be removed and just be a local variable.
However this is unrelated to the changes in this PR. Let's save this for   a separate PR

File: base/interactiveutil.jl [hunk 4]
```diff
@@ -571,11 +571,10 @@ end
...
 if is_windows()
     function download(url::AbstractString, filename::AbstractString)
-        res = ccall((:URLDownloadToFileW,:urlmon),stdcall,Cuint,
-                    (Ptr{Void},Cwstring,Cwstring,Cuint,Ptr{Void}),C_NULL,url,filename,0,C_NULL)
-        if res != 0
-            error("automatic download failed (error: $res): $url")
-        end
+        ps = "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe"
+        enable_tls12 = "[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.SecurityProtocolType]::Tls12"
+        download_cmd = "(New-Object System.Net.Webclient).DownloadFile('$(replace(url, "'", "''"))', '$(replace(filename, "'", "''"))')"
```
Comment (tkelman) [inline base/interactiveutil.jl]: this replacement doesn't make a lot of sense, could use comments and tests to show what it's accomplishing and why it's necessary

Comment (musm) [inline base/interactiveutil.jl]: See comment at https://github.com/JuliaLang/julia/pull/21968#discussion_r117542725 . Not really sure why it's needed so someoneelse would need to add tests and comments for this

File: base/interactiveutil.jl (hunk 1 above)
Comment (tkelman) [inline base/interactiveutil.jl]: what is a counterexample and how does this fix it?

Comment (musm) [inline base/interactiveutil.jl]: @vtjnash mind providing some test cases for this since I don't understand why it's needed?

Comment (musm) [inline base/interactiveutil.jl]: @vtjnash  (mind providing some test cases for this since I don't understand why it's needed?)

File: base/interactiveutil.jl (hunk 4 above)
Comment (vtjnash) [inline base/interactiveutil.jl]: try fetching a URL stuffed with metacharacters like `https://httpbin.org/get?test='^"`

Comment (musm) [inline base/interactiveutil.jl]: I think I see your point, but don't you want 
`replace(filename, "'", "\'")`
instead of
`replace(filename, "'", "''")`
or are those the same

and is this only needed here since we first interpolate the string into another string before  interpolating into the cmd  `run(` `)` ? I'm just trying to understand why the non windows download cmd doesn't need this.

Comment (vtjnash) [inline base/interactiveutil.jl]: No, the escape sequence for `'` is `''`, not `\'`

Non-windows can call the download program directly, and doesn't need to involve a shell.

Comment (tkelman) [inline base/interactiveutil.jl]: you mean ``` `' ``` ?

Comment (vtjnash) [inline base/interactiveutil.jl]: no, the escape-character is context sensitive. I used single quotes because the rules are simplest (https://ss64.com/ps/syntax-esc.html)

**Comment by musm** [comment]
test added

File: base/interactiveutil.jl (hunk 3 above)
Comment (musm) [inline base/interactiveutil.jl]: how about moving the check outside so that we write here
```julia
const downloadcmd = for checkcmd in (:curl, :wget, :fetch)
    if success(pipeline(`which $checkcmd`, DevNull))
        downloadcmd = checkcmd
        break
    end
end


```
and remove that from inside the `download` function ?

Comment (musm) [inline base/interactiveutil.jl]: Not sure on a good way to avoid the global here?

File: base/interactiveutil.jl [hunk 5]
```diff
@@ -598,7 +598,7 @@ else
                 rethrow()
             end
         elseif downloadcmd == :curl
-            run(`curl -L -f -o $filename $url`)
+            run(`curl -L -f -s -o $filename $url`)
```
Comment (tkelman) [inline base/interactiveutil.jl]: separate PR please

Comment (musm) [inline base/interactiveutil.jl]: Will do

Comment (tkelman) [inline base/interactiveutil.jl]: you've seen https://github.com/JuliaLang/BinDeps.jl/issues/135, but this function could probably use a more serious refactoring - not everyone will want all downloads to be silent. you can use curlrc to control some of these globally if you know you're using curl

Comment (musm) [inline base/interactiveutil.jl]: Yeah I think that will require more serious thinking about the api etc.  I think for now we should at least try to make the experience consistent; dumping out the progress bar  from curl directly doesn't seem very Julian. I agree it would be useful to have some control over the progress etc.

**Comment by musm** [comment]
anyone mind retriggering travis ?

**Comment by nalimilan** [comment]
Done.

**Comment by KristofferC** [comment]
I plan on merging this tomorrow unless someone has an objection

**Comment by tkelman** [comment]
This should set the downloadcmd global.

File: base/interactiveutil.jl (hunk 3 above)
Comment (tkelman) [inline base/interactiveutil.jl]: you're changing the download function to shell out to something, the value of what you're shelling out to should be saved in the global that exists for that purpose

**Comment by musm** [comment]
The downloadcmd is only relevant for non windows systems. We don't set its value to anything currently.

**Comment by tkelman** [comment]
Now that the download function shells out on windows, that variable is relevant on windows

**Comment by musm** [comment]
How so? Looks to me that `downloadcmd` is used on linux to test which download shell function to use. On windows we only have `powershell` so I am a  bit confused as to your suggestion?

**Comment by tkelman** [comment]
set it to powershell on windows.

**Comment by musm** [comment]
Do we know what using powershell2 instead of `URLDownloadToFileW` solves?

**Comment by vtjnash** [comment]
The belief is that this will have better support for redirects, cookies, authentication, and file size among other issues

**Review by tkelman — APPROVED** [review]
needs rebase

File: base/interactiveutil.jl [hunk 6]
```diff
@@ -591,17 +591,19 @@ end
 
 # file downloading
 
-downloadcmd = nothing
```
Comment (tkelman) [inline base/interactiveutil.jl]: don't think this should be moved

**Comment by musm** [comment]
The only thing that is disappointing with shelling is that it is slower than the call to `URLDownloadToFileW` ....

**Comment by tkelman** [comment]
powershell has a kind of annoying first-time startup delay, hopefully isn't as bad if you download multiple things in a julia session

**Comment by musm** [comment]
>  powershell has a kind of annoying first-time startup delay, hopefully isn't as bad if you download multiple things in a julia session


Fair enough. Should we merge this then?

**Comment by musm** [comment]
bumping for decision

**Comment by musm** [comment]
It may be worth adding this as a non-functional change to `Compat` and using that as `Compat.download` in `BinDeps` and `WinRPM` , thoughts on this ?

**Comment by tkelman** [comment]
BinDeps already has been using powershell in its download_cmd

**Comment by musm** [comment]
Right, but at least then it would be using the same implementation (e.g. it doesn't look like it correctly escapes urls).

**Comment by musm** [comment]
@tkelman  Would you be on board with that? I.e. replace all uses of download with the definitions in base?

**Comment by tkelman** [comment]
The BinDeps version would have to be deprecated, as it's used by other packages

**Comment by musm** [comment]
Should we merge this or leave it as is ?

**Comment by musm** [comment]
I don't think this is important and causing issues

**Comment by nalimilan** [comment]
What kind of issues? I thought it was supposed to reduce download failures, which would be a major improvement. The PR has been approved for weeks, so why close it now?

**Comment by JeffBezanson** [comment]
Rebased: #25477
